src/DataSourceProvider.vue

```diff
@@ -158,7 +158,9 @@ export default {
         this.widgetData.accessRules.forEach((defaultRule, index, array) => {
           array[index].type = this.missingAccessTypes;
 
-          this.selectedDataSource.accessRules.push(defaultRule);
+          if (index === 0 || defaultRule.allow !== array[index - 1].allow) {
```

Comment:
I don't quite understand what the `allow` property is. Can you explain?

Reply:
@tonytlwu `allow` property is determinating to what group of the user's rule should be applied to. And now we are adding a new default rule only if it wasn't applied to the same user's group and with the same types.

Reply:
```js
this.widgetData.accessRules.forEach((rule) => { // @problem rule is fine
  // There's no need to use the index here
  rule.type = this.missingAccessTypes;
  rule.enabled = true;

  // If the same allow property can't be found in the existing rules, add the rule
})
```

Currently you're using `index - 1` which suggests that it's only looking for "the one before'.

Reply:
@tonytlwu  added suggestions